hathor/cli/side_dag.py

```diff
@@ -0,0 +1,210 @@
+class HathorProcessTerminated:
+    pass
+
+
+@dataclass(frozen=True, slots=True)
+class SideDagProcessTerminated:
+    pass
+
+
+class SideDag(RunNode):
+    env_vars_prefix = 'hathor_side_dag_'
+
+
+def main(capture_stdout: bool) -> None:
+    """
+    This command runs two full node instances in separate processes.
+
+    The main process runs a side-dag full node, and it accepts the same options as the `run_node` command. Options
+    with the `--side-dag` prefix are passed to the side-dag full node, while options without this prefix are passed
+    to the non-side-dag full node, which runs in a background process and is commonly just a Hathor full node.
+    Whenever one of the full nodes fail, the other is automatically terminated.
+
+    By default, both full nodes output logs to stdout, but logs can be configured independently. Here's an example:
+
+    ```bash
+    $ python -m hathor side_dag
+        --testnet
+        --procname-prefix testnet-
+        --memory-storage
+        --disable-logs
+        --side-dag-config-yaml ./my-side-dag.yml
+        --side-dag-procname-prefix my-side-dag-
+        --side-dag-memory-storage
+        --side-dag-json-logs
+    ```
+
+    In this example, Hathor testnet logs would be disabled, while side-dag logs would be outputted to stdout as json.
+    """
+    from hathor.cli.util import process_logging_options, setup_logging
+    hathor_node_argv, side_dag_argv = _process_argv(sys.argv[1:])
+    conn1, conn2 = Pipe()
+    hathor_node_process = _start_hathor_node_process(hathor_node_argv, capture_stdout=capture_stdout, conn=conn1)
+
+    debug, logging_output, sentry = process_logging_options(side_dag_argv)
+    setup_logging(debug=debug, capture_stdout=capture_stdout, logging_output=logging_output, sentry=sentry)
+    logger.info('starting nodes', hathor_node_pid=hathor_node_process.pid, side_dag_pid=os.getpid())
+
+    _run_side_dag_node(side_dag_argv, hathor_node_process=hathor_node_process, conn=conn2)
+
+
+def _process_argv(argv: list[str]) -> tuple[list[str], list[str]]:
+    """Partition arguments into hathor node args and side-dag args, based on the `--side-dag` prefix."""
+    hathor_node_argv: list[str] = []
+    side_dag_argv: list[str] = []
+
+    def is_option(arg_: str) -> bool:
+        return arg_.startswith('--')
+
+    for i, arg in enumerate(argv):
+        if not is_option(arg):
+            continue
+
+        try:
+            value = None if is_option(argv[i + 1]) else argv[i + 1]
+        except IndexError:
+            value = None
+
+        argv_list = hathor_node_argv
+        if arg.startswith('--side-dag'):
+            arg = arg.replace('--side-dag-', '--')
+            argv_list = side_dag_argv
+
+        argv_list.append(arg)
+        if value is not None:
+            argv_list.append(value)
+
+    return hathor_node_argv, side_dag_argv
+
+
+def _run_side_dag_node(argv: list[str], *, hathor_node_process: Process, conn: 'Connection') -> None:
```

Comment:
If I got it right, the current process will spawn a new process for the mainnet full node and run the side-dag full node in the current process. Is that right? Isn't it better to spawn two new processes, one for each full node and develop a process monitor that gracefully terminates when one of the processes dies or freezes?

About signals, all processes would have to handle signals.

Reply:
As discussed in person, let's do a monitor process in a separate PR so we can skip it for now.

Reply:
Created an issue for this: https://github.com/HathorNetwork/hathor-core/issues/1051